### PR TITLE
Cosmos DB: Migration to V3 SDK

### LINKED ISSUE #610: Update CosmosDB SDK
Hello folks!

Is there any plans to update the CosmosDB bindings and changefeed host on Azure Functions to use the new [CosmosDB library](https://www.nuget.org/packages/Microsoft.Azure.Cosmos/)?

Would that be an acceptable PR?

Thanks!

### PR DESCRIPTION
**Pull request by ealsur**
This PR migrates the current implementation to V3 SDK and bumps the version to 4.0.0.

Closes #610

### CODE REVIEW

**Comment by pragnagopa** [comment]
Tagging @divyagandhisethi - for reference - for PR: https://github.com/kedacore/keda-external-scaler-azure-cosmos-db/pull/1

File: src/WebJobs.Extensions.CosmosDB/Bindings/CosmosDBAsyncCollector.cs [hunk 1]
```diff
@@ -63,18 +56,16 @@ public Task FlushAsync(CancellationToken cancellationToken = default(Cancellatio
...
         {
-            Uri collectionUri = UriFactory.CreateDocumentCollectionUri(context.ResolvedAttribute.DatabaseName, context.ResolvedAttribute.CollectionName);
-
-            // DocumentClient does not accept strings directly.
-            object convertedItem = item;
+            // Support user sending a string
             if (item is string)
             {
-                convertedItem = JObject.Parse(item.ToString());
+                JObject asJObject = JObject.Parse(item.ToString());
```
Comment (brettsam) [inline src/WebJobs.Extensions.CosmosDB/Bindings/CosmosDBAsyncCollector.cs:64]: Can we use System.Text.Json here? Would that allow you to remove a ref to Newtonsoft?

File: src/WebJobs.Extensions.CosmosDB/Bindings/CosmosDBItemValueBinder.cs [hunk 2]
```diff
@@ -41,48 +40,35 @@ public async Task SetValueAsync(object value, CancellationToken cancellationToke
...
+            if (typeof(T) != typeof(string))
             {
-                _originalItem = JObject.FromObject(document);
-                item = _originalItem.ToString(Formatting.None) as T;
+                try
+                {
+                    document = await _context.Service.GetContainer(_context.ResolvedAttribute.DatabaseName, _context.ResolvedAttribute.CollectionName)
+                        .ReadItemAsync<T>(_context.ResolvedAttribute.Id, partitionKey);
+
+                    _originalItem = JObject.FromObject(document);
```
Comment (brettsam) [inline src/WebJobs.Extensions.CosmosDB/Bindings/CosmosDBItemValueBinder.cs:55]: I'm wondering if we want to continue supporting this "update-in-place" behavior. It's never worked for out-of-proc languages and has resulted in a decent amount of confusion as it seems all of the bindings handle it differently.

@fabiocav -- thoughts?

File: src/WebJobs.Extensions.CosmosDB/Config/CosmosDBOptions.cs [hunk 3]
```diff
@@ -17,70 +16,20 @@ public class CosmosDBOptions : IOptionsFormatter
...
 
-        /// <summary>
-        /// Gets or sets the Protocol used in the DocumentClient instances.
-        /// </summary>
-        public Protocol? Protocol { get; set; }
-
-        /// <summary>
-        /// Gets or sets the lease options for the DocumentDB Trigger. 
-        /// </summary>
-        public ChangeFeedHostOptions LeaseOptions { get; set; } = new ChangeFeedHostOptions();
```
Comment (brettsam) [inline src/WebJobs.Extensions.CosmosDB/Config/CosmosDBOptions.cs:32]: Are these not settable now?

File: src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerAttribute.cs [hunk 4]
```diff
@@ -39,139 +39,98 @@ public CosmosDBTriggerAttribute(string databaseName, string collectionName)
         }
 
         /// <summary>
-        /// Connection string for the service containing the collection to monitor
+        /// Gets or sets the connection string for the service containing the collection to monitor.
         /// </summary>
-        [AppSetting]
+        [ConnectionString]
```
Comment (brettsam) [inline src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerAttribute.cs:44]: I should probably know this... but what's the difference between this and `[AppSetting]`?

File: src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerAttributeBindingProviderGenerator.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            CosmosDBTriggerAttribute cosmosDBTriggerAttribute = context.Parameter.GetCustomAttribute<CosmosDBTriggerAttribute>(inherit: false);
+
+            if (cosmosDBTriggerAttribute == null)
+            {
+                return Task.FromResult<ITriggerBinding>(null);
+            }
+
+            Type documentType = CosmosDBTriggerAttributeBindingProviderGenerator.GetParameterType(context.Parameter);
+
+            if (typeof(JArray).IsAssignableFrom(documentType))
```
Comment (brettsam) [inline src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerAttributeBindingProviderGenerator.cs:48]: Oh right... I forgot we required a Newtonsoft ref...

File: src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerAttributeBindingProviderGenerator.cs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            {
+                documentType = typeof(JObject); // When binding to JArray, use JObject as contract.
+            }
+
+            // TODO: What's up with string
+            if (typeof(string).IsAssignableFrom(documentType))
+            {
+                documentType = typeof(JObject); // When binding to JArray, use JObject as contract.
```
Comment (brettsam) [inline src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerAttributeBindingProviderGenerator.cs]: Does this work?

File: src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerBinding.cs [hunk 7]
```diff
@@ -96,31 +96,20 @@ public ParameterDescriptor ToParameterDescriptor()
...
-            }
-            catch
+            if (value == null)
             {
-                return false;
+                throw new ArgumentNullException(nameof(value));
             }
+
+            Dictionary<string, object> bindingData = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
+            bindingData.Add("CosmosDBTrigger", value);
```
Comment (brettsam) [inline src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerBinding.cs]: Was this supported before? Not sure if we need this?

File: src/WebJobs.Extensions.CosmosDB/Bindings/CosmosDBAsyncCollector.cs (hunk 1 above)
Comment (brettsam) [reply src/WebJobs.Extensions.CosmosDB/Bindings/CosmosDBAsyncCollector.cs:64]: Nevermind... I forgot we require Newtonsoft.

**Review by brettsam — COMMENTED** [review]
Initial comments...

File: src/WebJobs.Extensions.CosmosDB/Bindings/CosmosDBAsyncCollector.cs (hunk 1 above)
Comment (ealsur) [reply src/WebJobs.Extensions.CosmosDB/Bindings/CosmosDBAsyncCollector.cs:64]: The SDK could take System.Text.Json using a custom serializer if that is the approach we want to take.

File: src/WebJobs.Extensions.CosmosDB/Config/CosmosDBOptions.cs (hunk 3 above)
Comment (ealsur) [reply src/WebJobs.Extensions.CosmosDB/Config/CosmosDBOptions.cs:32]: They were already there in the Trigger properties. Normally these are for a very advanced customization of the trigger, and having it as a general configuration (for all the Triggers in the project) wouldn't have sense.

File: src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerAttribute.cs (hunk 4 above)
Comment (ealsur) [reply src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerAttribute.cs:44]: I actually don't know, but the `CosmosDBAttribute` file had it as `ConnectionString` so thought it should be consistent?

File: src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerAttributeBindingProviderGenerator.cs (hunk 5 above)
Comment (ealsur) [reply src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerAttributeBindingProviderGenerator.cs:48]: If we were to change to System.Text.Json, which would be the equivalent?

File: src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerBinding.cs (hunk 7 above)
Comment (ealsur) [reply src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerBinding.cs]: I thought it was required, but I removed it and still worked.

File: src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerAttributeBindingProviderGenerator.cs (hunk 6 above)
Comment (ealsur) [reply src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerAttributeBindingProviderGenerator.cs]: Yes, I added E2E tests for it.

Comment (ealsur) [reply src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerAttributeBindingProviderGenerator.cs]:
```
            public static void TriggerWithString(
                [CosmosDBTrigger(DatabaseName, CollectionName, CreateLeaseCollectionIfNotExists = true, LeaseCollectionPrefix = "withstring")] string documents,
                ILogger log)
            {
                foreach (var document in JArray.Parse(documents))
                {
                    log.LogInformation("Trigger with string called!");
                }
            }
```

**Review by brettsam — APPROVED** [review]

**Pull request merged**